src/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/Base64Transforms.cs

```diff
@@ -96,143 +103,211 @@ public class FromBase64Transform : ICryptoTransform
 
-            byte[] transformBuffer = new byte[_inputIndex + effectiveCount];
-            Buffer.BlockCopy(_inputBuffer, 0, transformBuffer, 0, _inputIndex);
-            Buffer.BlockCopy(temp, 0, transformBuffer, _inputIndex, effectiveCount);
+            const byte padding = (byte)'=';
+            int len = tmpBuffer.Length;
 
-            _inputIndex = (effectiveCount + _inputIndex) % 4;
-            Buffer.BlockCopy(temp, effectiveCount - _inputIndex, _inputBuffer, 0, _inputIndex);
+            // In Base64 there are maximum 2 padding chars
+            if (tmpBuffer[len - 2] == padding) outputSize--;
+            if (tmpBuffer[len - 1] == padding) outputSize--;
 
-            char[] tempChar = Encoding.ASCII.GetChars(transformBuffer, 0, 4 * numBlocks);
-            byte[] tempBytes = Convert.FromBase64CharArray(tempChar, 0, 4 * numBlocks);
-            return tempBytes;
+            return outputSize;
         }
 
-        private byte[] DiscardWhiteSpaces(byte[] inputBuffer, int inputOffset, int inputCount)
+        private void ConvertFromBase64(Span<byte> tmpBuffer, Span<byte> outputBuffer, out int consumed, out int written)
         {
-            int i, iCount = 0;
-            for (i = 0; i < inputCount; i++)
+            int bytesToTransform = _inputIndex + tmpBuffer.Length;
+
+            // Common case for bytesToTransform = 4
+            byte[] transformBufferArray = null;
+            Span<byte> transformBuffer = stackalloc byte[4];
+            if (bytesToTransform > 4)
             {
-                if (char.IsWhiteSpace((char)inputBuffer[inputOffset + i])) iCount++;
+                transformBuffer = transformBufferArray = ArrayPool<byte>.Shared.Rent(bytesToTransform);
             }
 
-            // If there's nothing to do, leave early
-            if (iCount == 0 && inputOffset == 0 &&
-                inputCount == inputBuffer.Length)
+            transformBuffer = transformBuffer.Slice(0, bytesToTransform);
+
+            // Copy _inputBuffer to transformBuffer and append tmpBuffer
+            Debug.Assert(_inputIndex < _inputBuffer.Length);
+            _inputBuffer.AsSpan(0, _inputIndex).CopyTo(transformBuffer);
+            tmpBuffer.CopyTo(transformBuffer.Slice(_inputIndex));
+
+            // Save data that won't be transformed to _inputBuffer, so it can be transformed later
+            _inputIndex = bytesToTransform & 3;     // bit hack for % 4
+            Debug.Assert(_inputIndex < _inputBuffer.Length);
+            tmpBuffer.Slice(tmpBuffer.Length - _inputIndex).CopyTo(_inputBuffer);
+
+            OperationStatus status = Base64.DecodeFromUtf8(transformBuffer, outputBuffer, out consumed, out written);
+
+            if (status == OperationStatus.InvalidData)
             {
-                return inputBuffer;
+                ThrowHelper.ThrowBase64FormatException();
             }
+            Debug.Assert(status == OperationStatus.Done);
```

Comment:
Given this assert, it seems like the better defensive code (especially since this is in a crypto namespace) is

```C#
if (status != OperationStatus.Done)
{
    Debug.Assert(status == OperationStatus.InvalidData);
    ThrowHelper.ThrowBase64FormatException();
}
```